### PR TITLE
Make IES-implementation similar to paper

### LINKED ISSUE #3123: Make IES-implementation identical / as similar as possible to paper

### PR DESCRIPTION
**Pull request by dafeda**
**Issue**
Resolves #3123

## Pre review checklist

- [x] Added appropriate labels

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

File: libres/lib/analysis/ies/ies.cpp [hunk 1]
```diff
@@ -15,6 +15,13 @@
    See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
    for more details.
 */
+// Needed to make intellisense work with Eigen on arm.
+// https://github.com/microsoft/vscode-cpptools/issues/7413#issuecomment-827172897
+#if __INTELLISENSE__
```
Comment (dafeda) [inline libres/lib/analysis/ies/ies.cpp:20]: What is this voodoo @pinkwah ?

File: libres/lib/analysis/ies/ies.cpp [hunk 2]
```diff
@@ -84,37 +84,48 @@ Eigen::MatrixXd ies::makeX(const Eigen::MatrixXd &A, const Eigen::MatrixXd &Y0,
...
     /* Store previous W for convergence test */
     Eigen::MatrixXd W = W0;
 
     /*
      * COMPUTE NEW UPDATED W                                                                        (Line 9)
-     *    W = W + ies_steplength * ( W - S'*(S*S'+R)^{-1} H )          (a)
+     *    W = W - ies_steplength * ( W - S'*(S*S'+R)^{-1} H )          (a)
      * which in the case when R=I can be rewritten as
-     *    W = W + ies_steplength * ( W - (S'*S + I)^{-1} * S' * H )    (b)
+     *    W = W - ies_steplength * ( W - (S'*S + I)^{-1} * S' * H )    (b)
```
Comment (oyvindeide) [inline libres/lib/analysis/ies/ies.cpp:128]: This seems like a rather significant change, has this comment just always been wrong?

File: libres/lib/analysis/ies/ies.cpp [hunk 3]
```diff
@@ -138,27 +149,21 @@ Eigen::MatrixXd ies::makeX(const Eigen::MatrixXd &A, const Eigen::MatrixXd &Y0,
...
-     * CONSTRUCT TRANFORM MATRIX X FOR CURRENT ITERATION (Line 10)
-     *   X= I + W/sqrt(N-1)
-     */
+    /* Line 9 of Algorithm 1 */
     Eigen::MatrixXd X = W0;
     X /= sqrt(ens_size - 1.0);
     X.diagonal().array() += 1;
 
-    /* COMPUTE ||W0 - W|| AND EVALUATE COST FUNCTION FOR PREVIOUS ITERATE (Line 12)*/
-    Eigen::MatrixXd DW = W0 - W;
```
Comment (oyvindeide) [inline libres/lib/analysis/ies/ies.cpp:160]: Was this never used?

File: libres/lib/analysis/ies/ies.cpp (hunk 2 above)
Comment (dafeda) [reply libres/lib/analysis/ies/ies.cpp:128]: Plus minus, potato potato!

I think the comment has been wrong, see line 8 in the figure taken from the IES-paper.

Screenshot 2022-04-05 at 14 56 49

File: libres/lib/analysis/ies/ies.cpp (hunk 3 above)
Comment (dafeda) [reply libres/lib/analysis/ies/ies.cpp:160]: Looks like it was used when we wrote to files for use with Tecplot.
Nobody has complained yet, so I think it's ok to delete.

```cpp
matrix_type *DW = matrix_alloc(ens_size, ens_size);
matrix_sub(DW, W0, W);
teclog(W, D, DW, "iesteclog.dat", ens_size, iteration_nr, rcond, nrsing,
       nrobs_inp);
teccost(W, D, "costf.dat", ens_size, iteration_nr);
```

**Comment by ManInFez** [comment]
Test ert please!

File: libres/lib/analysis/ies/ies.cpp (hunk 2 above)
Comment (pinkwah) [reply libres/lib/analysis/ies/ies.cpp:128]: Are you sure it's not a typo in the figure?

Comment (dafeda) [reply libres/lib/analysis/ies/ies.cpp:128]: Do you think there's a typo in the figure?

**Comment by BjarneHerland** [comment]
The error you see in `ERT - pytest` might be fixed by https://github.com/equinor/ert/pull/3237. You may try to rebase and resubmit...

File: libres/lib/analysis/ies/ies.cpp [hunk 4]
```diff
@@ -15,6 +15,7 @@
    See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
    for more details.
 */
+
```
Comment (oyvindeide) [inline libres/lib/analysis/ies/ies.cpp:18]: Extra whitespace?

File: libres/lib/include/ert/analysis/utils.hpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+std::vector<int>
+bool_vector_to_active_list(const std::vector<bool> &bool_vector) {
+    std::vector<int> active_list;
+    for (int i = 0; i < bool_vector.size(); i++) {
+        if (bool_vector[i])
+            active_list.push_back(i);
+    }
+    return active_list;
+}
+} // namespace analysis
```
Comment (oyvindeide) [inline libres/lib/include/ert/analysis/utils.hpp:13]: missing newline

**Review by oyvindeide — APPROVED** [review]
LGTM!

**Pull request merged**